### PR TITLE
You can now request YouTube video to play!

### PR DESCRIPTION
**Pull request by eartle**
If anyone has ideas on how to get artist and title of a youtube music video, please let me know. Would mean we get track info and scrobbles for requests.

There's this node module we could use: [get-artist-title](https://www.npmjs.com/package/get-artist-title), but I think it just some string matching to split the youtube title. Would be nice to fetch it from somewhere that knows better.

### CODE REVIEW

**Comment by eartle** [comment]
Please have a look @jammus @hugovk

File: package.json [hunk 1]
```diff
@@ -16,10 +16,10 @@
     "http-proxy": "1.13.2",
     "argparse": "1.0.6",
     "cli-interact": "0.1.9",
-    "throttle": "",
-    "ytdl-core": "",
-    "fluent-ffmpeg": "",
-    "through2": ""
+    "throttle": "1.0.3",
+    "ytdl-core": "0.11.2",
```
Comment (eartle) [inline package.json]: Got caught out by not fixing dependencies and streaming broke. After messing about for a while I found that the latest version of ytdl-core was the cause. Not sure why yet, but downgrading it fixes it so I'm happy.

File: static/js/controller/YouTubeSearchController.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  $scope.update = _.debounce(function() {
+    $scope.choosenTrack = null;
+    $.get(
+      "https://www.googleapis.com/youtube/v3/search",
+      {
+        q: $scope.searchTerm,
+        part: "snippet",
+        maxResults: 25,
+        type: "video",
+        videoCategoryId: "10", // Music
```
Comment (eartle) [inline static/js/controller/YouTubeSearchController.js]: Perhaps there are other restrictions we should set. The docs on the API console seem good. Haven't read through all of them yet though: https://developers.google.com/apis-explorer/#p/youtube/v3/youtube.search.list?q=surfing&maxResults=25&part=snippet

**Comment by hugovk** [comment]
I think string matching is the best you can do.

I have [this Chrome extension scrobbler](https://github.com/david-sabata/web-scrobbler) installed. It's MIT and [here's what it does](https://github.com/david-sabata/web-scrobbler/blob/f1cb8724793cd5d8be2075fb1a0dec8e5aa440f0/core/content/util.js#L9-L32) (and see its `splitArtistTrack` and `splitString` too):
```js
	/**
	 * Extract artist and track from Youtube video title.
	 * @param  {String} videoTitle Youtube video title
	 * @param  {Array} separators Array of separators
	 * @return {Object} Object contains artist and track fields
	 */
	processYoutubeVideoTitle(videoTitle) {
		if (!videoTitle) {
			return this.emptyArtistTrack;
		}

		// Remove [genre] from the beginning of the title
		let title = videoTitle.replace(/^\[[^\]]+\]\s*-*\s*/i, '');

		let { artist, track } = this.splitArtistTrack(title);
		if (artist === null && track === null) {
			// Look for Artist "Track"
			let artistTrack = title.match(/(.+?)\s"(.+?)"/);
			if (artistTrack) {
				artist = artistTrack[1];
				track = artistTrack[2];
			}
		}
		return { artist, track };
	},
```

It occasionally has the artist and title the wrong way round. One way around that would be to perhaps ask Last.fm something like if `(artist==string1 and track==string2)` has more scrobbles than `(artist==string2 and track==string1)`.

**Comment by eartle** [comment]
@hugovk I guess we can do whatever we are going to do to work out what the track artist and title is from the video title in the search dialog and display it to the user. Whoever is requesting the song can choose one that looks correct (i.e. don't request a video with a bad title)

**Comment by eartle** [comment]
Although I just tried the node module I suggested on my [new favourite song](https://www.youtube.com/watch?v=4zvoc_euQ-w) and it didn't work so now I don't know what to do.

**Comment by hugovk** [comment]
[Submit an issue](https://github.com/goto-bus-stop/get-artist-title) (or even a pull request!) that if it doesn't find something, add an extra case to look for "`<track> by <artist>`".

Or simply get a [new favourite song](https://www.youtube.com/watch?v=ZTmPb4LC1Zw).

**Comment by eartle** [comment]
I've just deployed this and is working p well \o/

**Comment by hugovk** [comment]
Hooray! (The CI broke though.)

**Comment by eartle** [comment]
Not really sure how to fix the tests. `util.js` requires some fancy new javascript things and they're not supported in PhantomJS yet so it's breaking the Karma test run thing.

**Comment by eartle** [comment]
BOOM! Pretty happy with this now. We're now passing all the frontend javascript through babel, which is nice because it fixes the tests, but the reason the tests were failing was because we were not supporting older browsers, which we are again now, maybe even more so. Also means the frontend javascript can be written in more modern javascript (arrow functions and etc) that'll get translated back.

**Pull request merged**